presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/EnvironmentUp.java

```diff
@@ -124,7 +131,8 @@ public void run()
                 return;
             }
 
-            sleepUntilInterrupted();
+            Collection<DockerContainer> containers = environment.getContainers();
+            waitUntil(() -> containers.stream().noneMatch(ContainerState::isRunning));
```

Comment:
When `noneMatch(ContainerState::isRunning)` you should exit with a non-zero exit code.
Also, you can skip the next log line (`log.info("Exiting, the containers will exit too");`)

Reply:
Changing the exit code is actually a bit tricky since this is being executed inside a Runnable. It's doable, just not straightforward without some refactoring. How badly do you want the non-zero exit code?

Reply:
Oh, we could just throw an exception when this predicate returns true, that'd get passed back up through the exception handler to the main thread properly.

Reply:
Not perfect, but perhaps better than just exiting cleanly. This is an erroneous situation after all.

Reply:
> This is an erroneous situation after all.

How this code handles successful situation when all tests are passing? Who breaks this loop?

Reply:
This code implements `env up`. Tests are run thru `test run` command (a separate class)